File: backend/app/schemas/workflow.py
from pydantic import BaseModel

from app.schemas.common import PaginationResponseBase, SelectOption


class WorkflowActionOption(BaseModel):
    action: str
    label: str


class WorkflowTaskRecord(BaseModel):
    id: int
    workflow_name: str
    business_module: str
    business_key: str
    title: str
    applicant_name: str
    current_handler: str
    current_node: str
    priority: str
    status: str
    created_at: str
    due_at: str
    form_summary: str
    latest_comment: str | None = None
    available_actions: list[WorkflowActionOption] = []
    process_definition_key: str | None = None
    process_definition_id: str | None = None
    process_instance_id: str | None = None
    execution_id: str | None = None
    task_definition_key: str | None = None


class WorkflowTaskUpsert(BaseModel):
    workflow_name: str
    business_module: str
    business_key: str
    title: str
    applicant_name: str
    current_handler: str
    current_node: str
    priority: str
    status: str
    created_at: str
    due_at: str
    form_summary: str
    latest_comment: str | None = None
    process_definition_key: str | None = None
    process_definition_id: str | None = None
    process_instance_id: str | None = None
    execution_id: str | None = None
    task_definition_key: str | None = None


class WorkflowTaskListResponse(PaginationResponseBase):
    items: list[WorkflowTaskRecord]


class WorkflowTaskActionRequest(BaseModel):
    action: str
    comment: str | None = None


class WorkflowTaskActionLog(BaseModel):
    operated_at: str
    operator_username: str
    operator_full_name: str
    action: str
    action_label: str
    from_node: str
    to_node: str | None = None
    result_status: str
    comment: str | None = None


class WorkflowTaskDetailResponse(BaseModel):
    task: WorkflowTaskRecord
    history: list[WorkflowTaskActionLog]


class WorkflowStats(BaseModel):
    todo_total: int
    in_progress_total: int
    approved_total: int
    rejected_total: int
    overdue_total: int


class WorkflowOptionsResponse(BaseModel):
    workflow_name_options: list[SelectOption]
    business_module_options: list[SelectOption]
    applicant_options: list[SelectOption]
    handler_options: list[SelectOption]
    current_node_options: list[SelectOption]
    priority_options: list[SelectOption]
    status_options: list[SelectOption]

File: frontend/src/api/workflow.ts
import type { PagedResponse, PaginationParams, SelectOption } from './common'
import http from './http'


export type WorkflowActionOption = {
  action: string
  label: string
}


export type WorkflowTaskRecord = {
  id: number
  workflow_name: string
  business_module: string
  business_key: string
  title: string
  applicant_name: string
  current_handler: string
  current_node: string
  priority: string
  status: string
  created_at: string
  due_at: string
  form_summary: string
  latest_comment?: string | null
  available_actions: WorkflowActionOption[]
  process_definition_key?: string | null
  process_definition_id?: string | null
  process_instance_id?: string | null
  execution_id?: string | null
  task_definition_key?: string | null
}


export type WorkflowTaskUpsert = Omit<
  WorkflowTaskRecord,
  'id' | 'available_actions' | 'process_definition_key' | 'process_definition_id' | 'process_instance_id' | 'execution_id' | 'task_definition_key'
>


export type WorkflowStats = {
  todo_total: number
  in_progress_total: number
  approved_total: number
  rejected_total: number
  overdue_total: number
}


export type WorkflowOptions = {
  workflow_name_options: SelectOption[]
  business_module_options: SelectOption[]
  applicant_options: SelectOption[]
  handler_options: SelectOption[]
  current_node_options: SelectOption[]
  priority_options: SelectOption[]
  status_options: SelectOption[]
}


export type WorkflowTaskActionLog = {
  operated_at: string
  operator_username: string
  operator_full_name: string
  action: string
  action_label: string
  from_node: string
  to_node?: string | null
  result_status: string
  comment?: string | null
}


export type WorkflowTaskDetailResponse = {
  task: WorkflowTaskRecord
  history: WorkflowTaskActionLog[]
}


export type WorkflowTaskActionRequest = {
  action: string
  comment?: string | null
}


export function getWorkflowStats() {
  return http.get<WorkflowStats>('/workflow/stats')
}


export function getWorkflowOptions() {
  return http.get<WorkflowOptions>('/workflow/options')
}


export function listWorkflowTasks(params?: PaginationParams & { keyword?: string; status?: string; module?: string }) {
  return http.get<PagedResponse<WorkflowTaskRecord>>('/workflow/tasks', { params })
}


export function getWorkflowTaskDetail(id: number) {
  return http.get<WorkflowTaskDetailResponse>(`/workflow/tasks/${id}`)
}


export function createWorkflowTask(payload: WorkflowTaskUpsert) {
  return http.post<WorkflowTaskRecord>('/workflow/tasks', payload)
}


export function updateWorkflowTask(id: number, payload: WorkflowTaskUpsert) {
  return http.put<WorkflowTaskRecord>(`/workflow/tasks/${id}`, payload)
}


export function deleteWorkflowTask(id: number) {
  return http.delete(`/workflow/tasks/${id}`)
}


export function executeWorkflowTaskAction(id: number, payload: WorkflowTaskActionRequest) {
  return http.post<WorkflowTaskDetailResponse>(`/workflow/tasks/${id}/actions`, payload)
}
